File: src/blackboard.test.ts
import { describe, it, expect } from 'vitest';
import { ScopedBlackboardReader, ScopedBlackboard } from './blackboard';
import { BlackboardEntry, BlackboardSource, BlackboardWrite } from './types';

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/** Create a BlackboardEntry with minimal required fields. */
function entry(
  key: string,
  value: unknown,
  opts: { workflowId?: string; nodeId?: string; stackDepth?: number; timestamp?: number } = {},
): BlackboardEntry {
  return {
    key,
    value,
    source: {
      workflowId: opts.workflowId ?? 'wf',
      nodeId: opts.nodeId ?? 'n',
      stackDepth: opts.stackDepth ?? 0,
    },
    timestamp: opts.timestamp ?? Date.now(),
  };
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

describe('ScopedBlackboardReader', () => {
  // -----------------------------------------------------------------------
  // Empty reader
  // -----------------------------------------------------------------------

  describe('empty reader', () => {
    it('returns undefined from get()', () => {
      const reader = new ScopedBlackboardReader([]);
      expect(reader.get('anything')).toBeUndefined();
    });

    it('returns false from has()', () => {
      const reader = new ScopedBlackboardReader([]);
      expect(reader.has('anything')).toBe(false);
    });

    it('returns empty array from getAll()', () => {
      const reader = new ScopedBlackboardReader([]);
      expect(reader.getAll('anything')).toEqual([]);
    });

    it('returns empty array from entries()', () => {
      const reader = new ScopedBlackboardReader([]);
      expect(reader.entries()).toEqual([]);
    });

    it('returns empty array from keys()', () => {
      const reader = new ScopedBlackboardReader([]);
      expect(reader.keys()).toEqual([]);
    });

    it('returns empty array from local()', () => {
      const reader = new ScopedBlackboardReader([]);
      expect(reader.local()).toEqual([]);
    });

    it('handles no-arg constructor as empty', () => {
      const reader = new ScopedBlackboardReader();
      expect(reader.get('x')).toBeUndefined();
      expect(reader.entries()).toEqual([]);
    });
  });

  // -----------------------------------------------------------------------
  // Single scope
  // -----------------------------------------------------------------------

  describe('single scope', () => {
    it('get() returns value for existing key', () => {
      const e = entry('color', 'blue');
      const reader = new ScopedBlackboardReader([[e]]);
      expect(reader.get('color')).toBe('blue');
    });

    it('has() returns true for existing key', () => {
      const reader = new ScopedBlackboardReader([[entry('color', 'blue')]]);
      expect(reader.has('color')).toBe(true);
    });

    it('has() returns false for missing key', () => {
      const reader = new ScopedBlackboardReader([[entry('color', 'blue')]]);
      expect(reader.has('size')).toBe(false);
    });

    it('keys() returns all unique keys', () => {
      const reader = new ScopedBlackboardReader([
        [entry('color', 'blue'), entry('size', 'large')],
      ]);
      expect(reader.keys()).toContain('color');
      expect(reader.keys()).toContain('size');
      expect(reader.keys()).toHaveLength(2);
    });

    it('get() returns each key correctly with multiple entries', () => {
      const reader = new ScopedBlackboardReader([
        [entry('color', 'blue'), entry('size', 'large')],
      ]);
      expect(reader.get('color')).toBe('blue');
      expect(reader.get('size')).toBe('large');
    });

    it('same-key shadowing: get() returns latest entry (last in array)', () => {
      const reader = new ScopedBlackboardReader([
        [entry('color', 'blue'), entry('color', 'red')],
      ]);
      expect(reader.get('color')).toBe('red');
    });

    it('same-key shadowing: getAll() returns both entries in order', () => {
      const e1 = entry('color', 'blue');
      const e2 = entry('color', 'red');
      const reader = new ScopedBlackboardReader([[e1, e2]]);
      const all = reader.getAll('color');
      expect(all).toHaveLength(2);
      expect(all[0].value).toBe('blue');
      expect(all[1].value).toBe('red');
    });

    it('local() returns all entries from the single scope', () => {
      const e1 = entry('color', 'blue');
      const e2 = entry('size', 'large');
      const reader = new ScopedBlackboardReader([[e1, e2]]);
      const loc = reader.local();
      expect(loc).toHaveLength(2);
      expect(loc[0]).toEqual(e1);
      expect(loc[1]).toEqual(e2);
    });
  });

  // -----------------------------------------------------------------------
  // Multi-scope: lexical scoping
  // -----------------------------------------------------------------------

  describe('multi-scope lexical scoping', () => {
    it('local shadows parent: get() returns local value', () => {
      const local = [entry('color', 'local-red')];
      const parent = [entry('color', 'parent-blue')];
      const reader = new ScopedBlackboardReader([local, parent]);
      expect(reader.get('color')).toBe('local-red');
    });

    it('parent fallback: get() returns parent value when not in local', () => {
      const local = [entry('size', 'large')];
      const parent = [entry('color', 'parent-blue')];
      const reader = new ScopedBlackboardReader([local, parent]);
      expect(reader.get('color')).toBe('parent-blue');
    });

    it('grandparent fallback: get() walks 3 scopes deep', () => {
      const local: BlackboardEntry[] = [];
      const parent: BlackboardEntry[] = [];
      const grandparent = [entry('origin', 'root-value')];
      const reader = new ScopedBlackboardReader([local, parent, grandparent]);
      expect(reader.get('origin')).toBe('root-value');
    });

    it('has() returns true if key in any scope', () => {
      const local: BlackboardEntry[] = [];
      const parent = [entry('color', 'blue')];
      const reader = new ScopedBlackboardReader([local, parent]);
      expect(reader.has('color')).toBe(true);
    });

    it('has() returns false if key in no scope', () => {
      const local = [entry('size', 'big')];
      const parent = [entry('color', 'blue')];
      const reader = new ScopedBlackboardReader([local, parent]);
      expect(reader.has('weight')).toBe(false);
    });
  });

  // -----------------------------------------------------------------------
  // getAll() across scopes
  // -----------------------------------------------------------------------

  describe('getAll() across scopes', () => {
    it('includes shadowed entries from all scopes, most-local first', () => {
      const local = [entry('color', 'local-red')];
      const parent = [entry('color', 'parent-blue')];
      const grandparent = [entry('color', 'gp-green')];
      const reader = new ScopedBlackboardReader([local, parent, grandparent]);

      const all = reader.getAll('color');
      expect(all).toHaveLength(3);
      expect(all[0].value).toBe('local-red');
      expect(all[1].value).toBe('parent-blue');
      expect(all[2].value).toBe('gp-green');
    });

    it('preserves entry order within each scope', () => {
      const local = [entry('x', 1), entry('x', 2)];
      const parent = [entry('x', 10), entry('x', 20)];
      const reader = new ScopedBlackboardReader([local, parent]);

      const all = reader.getAll('x');
      expect(all).toHaveLength(4);
      expect(all.map((e) => e.value)).toEqual([1, 2, 10, 20]);
    });

    it('returns empty for non-existent key', () => {
      const reader = new ScopedBlackboardReader([[entry('a', 1)]]);
      expect(reader.getAll('z')).toEqual([]);
    });
  });

  // -----------------------------------------------------------------------
  // entries() across scopes
  // -----------------------------------------------------------------------

  describe('entries()', () => {
    it('concatenates all scopes, local first', () => {
      const eLocal = entry('a', 1);
      const eParent = entry('b', 2);
      const eGP = entry('c', 3);
      const reader = new ScopedBlackboardReader([[eLocal], [eParent], [eGP]]);

      const all = reader.entries();
      expect(all).toHaveLength(3);
      expect(all[0]).toEqual(eLocal);
      expect(all[1]).toEqual(eParent);
      expect(all[2]).toEqual(eGP);
    });
  });

  // -----------------------------------------------------------------------
  // keys() deduplication
  // -----------------------------------------------------------------------

  describe('keys()', () => {
    it('deduplicates keys across scopes', () => {
      const local = [entry('color', 'red')];
      const parent = [entry('color', 'blue'), entry('size', 'big')];
      const reader = new ScopedBlackboardReader([local, parent]);

      const k = reader.keys();
      expect(k).toHaveLength(2);
      expect(k).toContain('color');
      expect(k).toContain('size');
    });
  });

  // -----------------------------------------------------------------------
  // local()
  // -----------------------------------------------------------------------

  describe('local()', () => {
    it('returns only innermost scope entries', () => {
      const eLocal = entry('a', 1);
      const eParent = entry('b', 2);
      const reader = new ScopedBlackboardReader([[eLocal], [eParent]]);

      const loc = reader.local();
      expect(loc).toHaveLength(1);
      expect(loc[0]).toEqual(eLocal);
    });

    it('returns empty array when innermost scope is empty', () => {
      const reader = new ScopedBlackboardReader([[], [entry('b', 2)]]);
      expect(reader.local()).toEqual([]);
    });

    it('returns a copy — mutations do not affect reader', () => {
      const e = entry('a', 1);
      const reader = new ScopedBlackboardReader([[e]]);
      const loc = reader.local();
      loc.pop();
      expect(reader.local()).toHaveLength(1);
    });
  });
});

// ===========================================================================
// ScopedBlackboard (write side)
// ===========================================================================

describe('ScopedBlackboard', () => {
  const source: BlackboardSource = {
    workflowId: 'wf-1',
    nodeId: 'node-A',
    stackDepth: 0,
  };

  // -----------------------------------------------------------------------
  // append() basics
  // -----------------------------------------------------------------------

  describe('append()', () => {
    it('creates entries with correct key and value', () => {
      const bb = new ScopedBlackboard();
      bb.append([{ key: 'color', value: 'blue' }], source);
      const entries = bb.getEntries();
      expect(entries).toHaveLength(1);
      expect(entries[0].key).toBe('color');
      expect(entries[0].value).toBe('blue');
    });

    it('attaches source metadata to every entry', () => {
      const bb = new ScopedBlackboard();
      bb.append([{ key: 'x', value: 1 }], source);
      const e = bb.getEntries()[0];
      expect(e.source.workflowId).toBe('wf-1');
      expect(e.source.nodeId).toBe('node-A');
      expect(e.source.stackDepth).toBe(0);
    });

    it('attaches a numeric timestamp to every entry', () => {
      const bb = new ScopedBlackboard();
      const before = Date.now();
      bb.append([{ key: 'x', value: 1 }], source);
      const after = Date.now();
      const ts = bb.getEntries()[0].timestamp;
      expect(ts).toBeGreaterThanOrEqual(before);
      expect(ts).toBeLessThanOrEqual(after);
    });

    it('gives all entries in one call the same timestamp and source', () => {
      const bb = new ScopedBlackboard();
      const writes: BlackboardWrite[] = [
        { key: 'a', value: 1 },
        { key: 'b', value: 2 },
        { key: 'c', value: 3 },
      ];
      bb.append(writes, source);
      const entries = bb.getEntries();
      expect(entries).toHaveLength(3);
      const ts = entries[0].timestamp;
      for (const e of entries) {
        expect(e.timestamp).toBe(ts);
        expect(e.source).toBe(source);
      }
    });

    it('returns the newly created entries', () => {
      const bb = new ScopedBlackboard();
      const result = bb.append([{ key: 'x', value: 42 }], source);
      expect(result).toHaveLength(1);
      expect(result[0].key).toBe('x');
      expect(result[0].value).toBe(42);
      expect(result[0].source).toBe(source);
    });
  });

  // -----------------------------------------------------------------------
  // Append-only invariant
  // -----------------------------------------------------------------------

  describe('append-only invariant', () => {
    it('entries accumulate across multiple append() calls', () => {
      const bb = new ScopedBlackboard();
      bb.append([{ key: 'a', value: 1 }], source);
      bb.append([{ key: 'b', value: 2 }], source);
      bb.append([{ key: 'c', value: 3 }], source);
      expect(bb.getEntries()).toHaveLength(3);
    });

    it('existing entries are unchanged after a new append', () => {
      const bb = new ScopedBlackboard();
      bb.append([{ key: 'a', value: 1 }], source);
      const first = bb.getEntries()[0];
      bb.append([{ key: 'b', value: 2 }], source);
      const entries = bb.getEntries();
      expect(entries[0].key).toBe(first.key);
      expect(entries[0].value).toBe(first.value);
      expect(entries[0].timestamp).toBe(first.timestamp);
    });
  });

  // -----------------------------------------------------------------------
  // Same-key shadowing (via reader integration)
  // -----------------------------------------------------------------------

  describe('same-key shadowing', () => {
    it('reader.get() returns the latest value for a shadowed key', () => {
      const bb = new ScopedBlackboard();
      bb.append([{ key: 'color', value: 'blue' }], source);
      bb.append([{ key: 'color', value: 'red' }], source);
      const reader = bb.reader();
      expect(reader.get('color')).toBe('red');
    });
  });

  // -----------------------------------------------------------------------
  // getEntries()
  // -----------------------------------------------------------------------

  describe('getEntries()', () => {
    it('returns current entries reflecting all appends', () => {
      const bb = new ScopedBlackboard();
      bb.append([{ key: 'x', value: 1 }], source);
      bb.append([{ key: 'y', value: 2 }], source);
      const entries = bb.getEntries();
      expect(entries).toHaveLength(2);
      expect(entries[0].key).toBe('x');
      expect(entries[1].key).toBe('y');
    });

    it('returns a copy — mutations do not affect internal state', () => {
      const bb = new ScopedBlackboard();
      bb.append([{ key: 'x', value: 1 }], source);
      const snapshot = bb.getEntries();
      // Mutate the returned array
      (snapshot as BlackboardEntry[]).push(entry('hack', 99));
      expect(bb.getEntries()).toHaveLength(1);
    });
  });

  // -----------------------------------------------------------------------
  // reader()
  // -----------------------------------------------------------------------

  describe('reader()', () => {
    it('constructs a working ScopedBlackboardReader', () => {
      const bb = new ScopedBlackboard();
      bb.append([{ key: 'color', value: 'green' }], source);
      const reader = bb.reader();
      expect(reader.get('color')).toBe('green');
      expect(reader.has('color')).toBe(true);
      expect(reader.keys()).toContain('color');
    });

    it('includes parent scopes in the reader', () => {
      const bb = new ScopedBlackboard();
      const parentEntries = [entry('origin', 'parent-value')];
      const reader = bb.reader([parentEntries]);
      expect(reader.get('origin')).toBe('parent-value');
    });

    it('local scope reflects writes, shadowing parent values', () => {
      const bb = new ScopedBlackboard();
      const parentEntries = [entry('color', 'parent-blue')];
      bb.append([{ key: 'color', value: 'local-red' }], source);
      const reader = bb.reader([parentEntries]);
      expect(reader.get('color')).toBe('local-red');
    });
  });

  // -----------------------------------------------------------------------
  // Empty blackboard
  // -----------------------------------------------------------------------

  describe('empty blackboard', () => {
    it('getEntries() returns empty array', () => {
      const bb = new ScopedBlackboard();
      expect(bb.getEntries()).toEqual([]);
    });

    it('reader().get() returns undefined', () => {
      const bb = new ScopedBlackboard();
      expect(bb.reader().get('anything')).toBeUndefined();
    });
  });

  // -----------------------------------------------------------------------
  // Constructor with seed entries
  // -----------------------------------------------------------------------

  describe('constructor with seed entries', () => {
    it('initializes with provided entries', () => {
      const seed = [entry('x', 1), entry('y', 2)];
      const bb = new ScopedBlackboard(seed);
      expect(bb.getEntries()).toHaveLength(2);
      expect(bb.reader().get('x')).toBe(1);
      expect(bb.reader().get('y')).toBe(2);
    });

    it('append adds to seed entries', () => {
      const seed = [entry('x', 1)];
      const bb = new ScopedBlackboard(seed);
      bb.append([{ key: 'y', value: 2 }], source);
      expect(bb.getEntries()).toHaveLength(2);
    });
  });
});

File: go/blackboard_test.go
package reflex

import (
	"sync"
	"testing"
)

// helper to create a BlackboardEntry
func bbEntry(key string, value any) BlackboardEntry {
	return BlackboardEntry{Key: key, Value: value, Source: BlackboardSource{WorkflowID: "wf", NodeID: "n", StackDepth: 0}, Timestamp: 1}
}

func bbEntryWithSource(key string, value any, wfID, nodeID string, depth int) BlackboardEntry {
	return BlackboardEntry{Key: key, Value: value, Source: BlackboardSource{WorkflowID: wfID, NodeID: nodeID, StackDepth: depth}, Timestamp: 1}
}

// ---------------------------------------------------------------------------
// ScopedBlackboardReader — empty
// ---------------------------------------------------------------------------

func TestBlackboardReaderEmpty(t *testing.T) {
	reader := NewBlackboardReader(nil)

	t.Run("Get returns false", func(t *testing.T) {
		_, ok := reader.Get("anything")
		if ok {
			t.Error("expected ok=false for empty reader")
		}
	})
	t.Run("Has returns false", func(t *testing.T) {
		if reader.Has("anything") {
			t.Error("expected Has=false")
		}
	})
	t.Run("GetAll returns empty", func(t *testing.T) {
		if len(reader.GetAll("anything")) != 0 {
			t.Error("expected empty")
		}
	})
	t.Run("Entries returns empty", func(t *testing.T) {
		if len(reader.Entries()) != 0 {
			t.Error("expected empty")
		}
	})
	t.Run("Keys returns empty", func(t *testing.T) {
		if len(reader.Keys()) != 0 {
			t.Error("expected empty")
		}
	})
	t.Run("Local returns empty", func(t *testing.T) {
		if len(reader.Local()) != 0 {
			t.Error("expected empty")
		}
	})
}

// ---------------------------------------------------------------------------
// ScopedBlackboardReader — single scope
// ---------------------------------------------------------------------------

func TestBlackboardReaderSingleScope(t *testing.T) {
	scope := []BlackboardEntry{bbEntry("color", "blue"), bbEntry("size", "large")}
	reader := NewBlackboardReader([][]BlackboardEntry{scope})

	t.Run("Get returns value for existing key", func(t *testing.T) {
		v, ok := reader.Get("color")
		if !ok || v != "blue" {
			t.Errorf("expected blue, got %v (ok=%v)", v, ok)
		}
	})
	t.Run("Has returns true for existing key", func(t *testing.T) {
		if !reader.Has("color") {
			t.Error("expected Has=true")
		}
	})
	t.Run("Has returns false for missing key", func(t *testing.T) {
		if reader.Has("weight") {
			t.Error("expected Has=false")
		}
	})
	t.Run("Keys returns all unique keys", func(t *testing.T) {
		keys := reader.Keys()
		if len(keys) != 2 {
			t.Errorf("expected 2 keys, got %d", len(keys))
		}
	})
	t.Run("Local returns scope entries", func(t *testing.T) {
		if len(reader.Local()) != 2 {
			t.Errorf("expected 2 local entries, got %d", len(reader.Local()))
		}
	})
	t.Run("Entries returns all entries", func(t *testing.T) {
		if len(reader.Entries()) != 2 {
			t.Errorf("expected 2 entries, got %d", len(reader.Entries()))
		}
	})
}

// ---------------------------------------------------------------------------
// ScopedBlackboardReader — shadowing within scope (latest wins)
// ---------------------------------------------------------------------------

func TestBlackboardReaderShadowingWithinScope(t *testing.T) {
	scope := []BlackboardEntry{bbEntry("color", "blue"), bbEntry("color", "red")}
	reader := NewBlackboardReader([][]BlackboardEntry{scope})

	t.Run("Get returns latest value", func(t *testing.T) {
		v, ok := reader.Get("color")
		if !ok || v != "red" {
			t.Errorf("expected red (latest), got %v", v)
		}
	})
	t.Run("GetAll returns full history", func(t *testing.T) {
		all := reader.GetAll("color")
		if len(all) != 2 {
			t.Errorf("expected 2 entries, got %d", len(all))
		}
	})
	t.Run("Keys deduplicates", func(t *testing.T) {
		keys := reader.Keys()
		if len(keys) != 1 {
			t.Errorf("expected 1 key, got %d", len(keys))
		}
	})
}

// ---------------------------------------------------------------------------
// ScopedBlackboardReader — multi-scope (child shadows parent)
// ---------------------------------------------------------------------------

func TestBlackboardReaderMultiScope(t *testing.T) {
	child := []BlackboardEntry{bbEntry("color", "red")}
	parent := []BlackboardEntry{bbEntry("color", "blue"), bbEntry("size", "large")}
	reader := NewBlackboardReader([][]BlackboardEntry{child, parent})

	t.Run("Get returns child value (shadows parent)", func(t *testing.T) {
		v, ok := reader.Get("color")
		if !ok || v != "red" {
			t.Errorf("expected red (child), got %v", v)
		}
	})
	t.Run("Get returns parent value for non-shadowed key", func(t *testing.T) {
		v, ok := reader.Get("size")
		if !ok || v != "large" {
			t.Errorf("expected large (parent), got %v", v)
		}
	})
	t.Run("Has finds key in any scope", func(t *testing.T) {
		if !reader.Has("size") {
			t.Error("expected Has=true for parent key")
		}
	})
	t.Run("GetAll returns entries across scopes local-first", func(t *testing.T) {
		all := reader.GetAll("color")
		if len(all) != 2 {
			t.Errorf("expected 2, got %d", len(all))
		}
		if all[0].Value != "red" {
			t.Error("expected local entry first")
		}
	})
	t.Run("Keys includes keys from all scopes", func(t *testing.T) {
		keys := reader.Keys()
		if len(keys) != 2 {
			t.Errorf("expected 2 keys (color, size), got %d", len(keys))
		}
	})
	t.Run("Local returns only child entries", func(t *testing.T) {
		local := reader.Local()
		if len(local) != 1 || local[0].Value != "red" {
			t.Error("expected only child entries")
		}
	})
}

// ---------------------------------------------------------------------------
// ScopedBlackboardReader — three scopes (grandparent)
// ---------------------------------------------------------------------------

func TestBlackboardReaderThreeScopes(t *testing.T) {
	grandchild := []BlackboardEntry{bbEntry("a", 1)}
	child := []BlackboardEntry{bbEntry("b", 2)}
	parent := []BlackboardEntry{bbEntry("c", 3)}
	reader := NewBlackboardReader([][]BlackboardEntry{grandchild, child, parent})

	t.Run("reads from each scope level", func(t *testing.T) {
		v, _ := reader.Get("a")
		if v != 1 {
			t.Errorf("expected 1, got %v", v)
		}
		v, _ = reader.Get("b")
		if v != 2 {
			t.Errorf("expected 2, got %v", v)
		}
		v, _ = reader.Get("c")
		if v != 3 {
			t.Errorf("expected 3, got %v", v)
		}
	})
	t.Run("Entries returns all in scope order", func(t *testing.T) {
		entries := reader.Entries()
		if len(entries) != 3 {
			t.Errorf("expected 3, got %d", len(entries))
		}
	})
}

// ---------------------------------------------------------------------------
// ScopedBlackboard — write side
// ---------------------------------------------------------------------------

func TestBlackboardAppend(t *testing.T) {
	bb := NewBlackboard()
	source := BlackboardSource{WorkflowID: "wf1", NodeID: "n1", StackDepth: 0}

	t.Run("Append returns new entries", func(t *testing.T) {
		writes := []BlackboardWrite{{Key: "x", Value: 42}}
		newEntries := bb.Append(writes, source)
		if len(newEntries) != 1 {
			t.Fatalf("expected 1 entry, got %d", len(newEntries))
		}
		if newEntries[0].Key != "x" || newEntries[0].Value != 42 {
			t.Error("entry mismatch")
		}
	})
	t.Run("Entries grows after append", func(t *testing.T) {
		bb.Append([]BlackboardWrite{{Key: "y", Value: 99}}, source)
		entries := bb.Entries()
		if len(entries) != 2 {
			t.Errorf("expected 2, got %d", len(entries))
		}
	})
	t.Run("Source is preserved", func(t *testing.T) {
		entries := bb.Entries()
		if entries[0].Source.WorkflowID != "wf1" {
			t.Error("source not preserved")
		}
	})
	t.Run("Reader with parent scopes", func(t *testing.T) {
		parentEntries := []BlackboardEntry{bbEntry("parent_key", "parent_val")}
		reader := bb.Reader(parentEntries)
		v, ok := reader.Get("parent_key")
		if !ok || v != "parent_val" {
			t.Error("expected parent key visible via reader")
		}
		v, ok = reader.Get("x")
		if !ok || v != 42 {
			t.Error("expected local key visible via reader")
		}
	})
}

// ---------------------------------------------------------------------------
// ScopedBlackboard — initial entries
// ---------------------------------------------------------------------------

func TestBlackboardWithInitialEntries(t *testing.T) {
	initial := []BlackboardEntry{bbEntry("init", "value")}
	bb := NewBlackboard(initial...)
	entries := bb.Entries()
	if len(entries) != 1 || entries[0].Key != "init" {
		t.Error("initial entries not preserved")
	}
}

// ---------------------------------------------------------------------------
// Cursor API
// ---------------------------------------------------------------------------

func TestCursorStartsAtZero(t *testing.T) {
	bb := NewBlackboard()
	if c := bb.Cursor(); c != 0 {
		t.Errorf("expected cursor 0, got %d", c)
	}
}

func TestCursorAdvancesWithAppends(t *testing.T) {
	bb := NewBlackboard()
	source := BlackboardSource{WorkflowID: "wf", NodeID: "n", StackDepth: 0}
	bb.Append([]BlackboardWrite{{Key: "a", Value: 1}, {Key: "b", Value: 2}, {Key: "c", Value: 3}}, source)
	if c := bb.Cursor(); c != 3 {
		t.Errorf("expected cursor 3, got %d", c)
	}
}

func TestEntriesFromZeroReturnsAll(t *testing.T) {
	bb := NewBlackboard()
	source := BlackboardSource{WorkflowID: "wf", NodeID: "n", StackDepth: 0}
	for i := 0; i < 5; i++ {
		bb.Append([]BlackboardWrite{{Key: "k", Value: i}}, source)
	}
	entries, next := bb.EntriesFrom(0)
	if len(entries) != 5 {
		t.Errorf("expected 5 entries, got %d", len(entries))
	}
	if next != 5 {
		t.Errorf("expected next cursor 5, got %d", next)
	}
}

func TestEntriesFromCursorReturnsDelta(t *testing.T) {
	bb := NewBlackboard()
	source := BlackboardSource{WorkflowID: "wf", NodeID: "n", StackDepth: 0}
	bb.Append([]BlackboardWrite{{Key: "a", Value: 1}, {Key: "b", Value: 2}, {Key: "c", Value: 3}}, source)
	cur := bb.Cursor()
	bb.Append([]BlackboardWrite{{Key: "d", Value: 4}, {Key: "e", Value: 5}}, source)

	entries, next := bb.EntriesFrom(cur)
	if len(entries) != 2 {
		t.Errorf("expected 2 delta entries, got %d", len(entries))
	}
	if next != 5 {
		t.Errorf("expected next cursor 5, got %d", next)
	}
	if entries[0].Key != "d" || entries[1].Key != "e" {
		t.Error("delta entries have wrong keys")
	}
}

func TestEntriesFromPastEnd(t *testing.T) {
	bb := NewBlackboard()
	source := BlackboardSource{WorkflowID: "wf", NodeID: "n", StackDepth: 0}
	bb.Append([]BlackboardWrite{{Key: "a", Value: 1}}, source)

	entries, next := bb.EntriesFrom(Cursor(999))
	if entries != nil {
		t.Errorf("expected nil entries, got %d", len(entries))
	}
	if next != 1 {
		t.Errorf("expected next cursor 1, got %d", next)
	}
}

func TestEntriesFromNegative(t *testing.T) {
	bb := NewBlackboard()
	source := BlackboardSource{WorkflowID: "wf", NodeID: "n", StackDepth: 0}
	bb.Append([]BlackboardWrite{{Key: "a", Value: 1}, {Key: "b", Value: 2}}, source)

	entries, next := bb.EntriesFrom(Cursor(-1))
	if len(entries) != 2 {
		t.Errorf("expected 2 entries, got %d", len(entries))
	}
	if next != 2 {
		t.Errorf("expected next cursor 2, got %d", next)
	}
}

func TestCursorConcurrentSafety(t *testing.T) {
	bb := NewBlackboard()
	source := BlackboardSource{WorkflowID: "wf", NodeID: "n", StackDepth: 0}

	var wg sync.WaitGroup
	// Writer goroutine
	wg.Add(1)
	go func() {
		defer wg.Done()
		for i := 0; i < 100; i++ {
			bb.Append([]BlackboardWrite{{Key: "k", Value: i}}, source)
		}
	}()
	// Reader goroutine using cursor
	wg.Add(1)
	go func() {
		defer wg.Done()
		var cur Cursor
		total := 0
		for total < 100 {
			entries, next := bb.EntriesFrom(cur)
			total += len(entries)
			cur = next
		}
	}()
	wg.Wait()
	if bb.Cursor() != 100 {
		t.Errorf("expected cursor 100, got %d", bb.Cursor())
	}
}

func TestCursorSeededBlackboard(t *testing.T) {
	seed := []BlackboardEntry{bbEntry("s1", "v1"), bbEntry("s2", "v2")}
	bb := NewBlackboard(seed...)

	if c := bb.Cursor(); c != 2 {
		t.Errorf("expected cursor 2 after seed, got %d", c)
	}
	entries, next := bb.EntriesFrom(0)
	if len(entries) != 2 {
		t.Errorf("expected 2 seed entries, got %d", len(entries))
	}
	if next != 2 {
		t.Errorf("expected next cursor 2, got %d", next)
	}
}

func TestCursorMultipleIncrementalReads(t *testing.T) {
	bb := NewBlackboard()
	source := BlackboardSource{WorkflowID: "wf", NodeID: "n", StackDepth: 0}

	var allRead []BlackboardEntry
	cur := bb.Cursor()

	// Batch 1
	bb.Append([]BlackboardWrite{{Key: "a", Value: 1}, {Key: "b", Value: 2}}, source)
	entries, cur := bb.EntriesFrom(cur)
	allRead = append(allRead, entries...)

	// Batch 2
	bb.Append([]BlackboardWrite{{Key: "c", Value: 3}}, source)
	entries, cur = bb.EntriesFrom(cur)
	allRead = append(allRead, entries...)

	// Batch 3
	bb.Append([]BlackboardWrite{{Key: "d", Value: 4}, {Key: "e", Value: 5}}, source)
	entries, _ = bb.EntriesFrom(cur)
	allRead = append(allRead, entries...)

	if len(allRead) != 5 {
		t.Errorf("expected 5 total entries (no duplicates), got %d", len(allRead))
	}
	expected := []string{"a", "b", "c", "d", "e"}
	for i, e := range allRead {
		if e.Key != expected[i] {
			t.Errorf("entry %d: expected key %s, got %s", i, expected[i], e.Key)
		}
	}
}

// ---------------------------------------------------------------------------
// Concurrent safety
// ---------------------------------------------------------------------------

func TestBlackboardConcurrentAppend(t *testing.T) {
	bb := NewBlackboard()
	source := BlackboardSource{WorkflowID: "wf", NodeID: "n", StackDepth: 0}
	var wg sync.WaitGroup
	for i := 0; i < 100; i++ {
		wg.Add(1)
		go func(i int) {
			defer wg.Done()
			bb.Append([]BlackboardWrite{{Key: "k", Value: i}}, source)
		}(i)
	}
	wg.Wait()
	if len(bb.Entries()) != 100 {
		t.Errorf("expected 100 entries, got %d", len(bb.Entries()))
	}
}
